### ISSUE TITLE
Empty BarChart shows "0 auto" as X Axis ticks

### ISSUE THREAD

**Issue by steven-schmoll-at** [recharts/recharts#2593]
- [x] I have searched the [issues](https://github.com/recharts/recharts/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[![Edit on CodeSandbox](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/simple-bar-chart-empty-x-axis-issue-tfep4)

### Steps to reproduce
Simply having a BarChart with an empty data array. Doesn't matter what other components are used within other than having the XAxis component. 

Just use any version of recharts 2.0.0 or higher including latest 2.0.10

### What is expected?
The chart will display with the 2 axis borders but no axis ticks.

### What is actually happening?
The chart displays 2 X Axis ticks:
- 0
- auto

| Environment | Info |
|---|---|
| Recharts | v2.0.8 |
| React | 17.0.1 |
| System | Linux |
| Browser | Firefox 89 or Chrome 91 |

---
This behavior has changed from version 1.X. Previously it did not display the X Axis at all if the data was empty.

See <https://codesandbox.io/s/recharts-issue-template-forked-45fdr> for example.

**Comment by bernardobelchior** [comment]
I'm experiencing the same issue on Recharts 2.1.6, macOS and Chrome and Firefox as well 👍

**Comment by bernardobelchior** [comment]
I'm now checking if the data array is empty and creating an empty label in order to avoid this problem. Something like this:

```js
const actualData = data.length === 0 ? [{ label: '' }] : data
```

**Comment by jessetabak** [comment]
I have the same issue in a LineChart (in recharts version 2.1.6)
It tries to use 'auto' (from the default domain) as a label for the tick.

In my case the 'auto' value is passed to the tickFormatter (on the XAxis)

As a workaround I just set a domain of [0, 10] if data.length === 0

**Comment by wingsbob** [comment]
Yeah, I've just hit this issue, except I've got a tick formatter which formats the date values on ticks, except `new Date('auto')` causes everything to explode :(

**Comment by jzeilmann93** [comment]
Hey, been dealing with this issue for a bit, try adding domain={['dataMin', 'dataMax']} to the X-Axis that is giving you an issue.

**Comment by svey** [comment]
i also have this issue any non hacks?

**Comment by ortonomy** [comment]
@svey and others - not sure it's a hack. Simply set `ticks={data.length ? undefined : [""]}` and it make a blank axis

**Comment by lukask-proxora** [comment]
After having a look at the code, it seems to me that the problem are the default props of the `XAxis` component. Those set the axis `type` to `category` and the `domain` to `[0, 'auto']`. This domain however only makes sense with axis type `number`. 
When determining the final domain in `generateCategoricalChart.tsx.getAxisMapByAxes` the domain prop of the axis is used if all values of `getDomainOfDataByKey` for this axis are in the defined domain. If the data array is empty `getDomainOfDataByKey` will return an empty array and thus all its values are contained in `[0, 'auto']`.

**Comment by lukask-proxora** [comment]
After having a look at the code, it seems to me that the problem are the default props of the `XAxis` component. Those set the axis `type` to `category` and the `domain` to `[0, 'auto']`. This domain however only makes sense with axis type `number`. 

When determining the final domain in `generateCategoricalChart.tsx.getAxisMapByAxes` the domain prop of the axis is used if all values of `getDomainOfDataByKey` for this axis are in the defined domain. If the data array is empty `getDomainOfDataByKey` will return an empty array and thus all its values are contained in `[0, 'auto']`. 

As a result, the problem also occurs if the data array contains one item with value 'auto' for the XAxis data key.

**Comment by lukask-proxora** [comment]
A simple workaround would be to set an empty domain:
`<XAxis dataKey="name" domain={[]} />`

This also works if the data array is not empty, since then the generated domain containing all name values in the data array would be used.

**Comment by ckifer** [comment]
Thanks for diving deep on this @lukask-proxora. This seems complicated with potential side-effects. 

I think you are onto something in the last statement you make - it doesn't have to be a default prop.

We can 
>Simply remove [0, 'auto'] from the default props of XAxis

and then we can set the "default" domain based on whats passed in rather than what is assigned as a default in the props.

Ultimately we would have a solution that "works" no matter what the domain is or types that limit the allowed domain based on the axis type

**Comment by lukask-proxora** [comment]
@ckifer Thanks for the help :+1: 

I was thinking about that before but thought that the different Axis components really could have different default domains. 
But I guess in general it makes sense for all Axis components to use `[0, 'auto']` as domain if the type is `number` and no domain is set on the axis and to use `[]` as domain if the type is `category` and no domain is set on the axis.

We could decide which default domain to take in the `generateCategoricalChart.tsx.getAxisMapByAxes` function then, was that your idea @ckifer ?

**Comment by ckifer** [comment]
Should be fixed in 2.4.0
